File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
   int subarraySum(vector<int>& nums, int k) {
       int n=nums.size();


       unordered_map<int,int>m;
       m[0]=1;
       int pre_sum=0;
       int ans=0;
       for(int i=0;i<n;i++)
       {
           pre_sum+=nums[i];
           if(m.find(pre_sum-k)!=m.end())
           ans+=m[pre_sum-k];


           m[pre_sum]++;
       }
       return ans;
   }
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        
        int n = nums.length;

        HashMap<Integer, Integer> map = new HashMap<>();
        map.put(0, 1); // initital prefix sum = 0 with 1 occurance

        int prefixSum = 0, ans = 0;

        for(int i = 0; i < n; i++) {
            prefixSum += nums[i];

            //check if (prefixSum - k) exists in the map
            if(map.containsKey(prefixSum - k)) {
                ans += map.get(prefixSum - k);
            }

            // store prefix sum frequency
            map.put(prefixSum, map.getOrDefault(prefixSum, 0) + 1);
        }

        return ans;
    }
}